test/distributed/_tensor/test_matrix_ops.py

```diff
@@ -120,6 +131,108 @@ def test_placement_comb(
         for spec in shard_specs_comb:
             test_placement_comb([spec[0]], [spec[1]])
 
+    @with_comms
+    @skip_unless_torch_gpu
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FP8, "torch._scaled_mm requires H100+")
+    def test_scaled_mm(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        shrd0 = Shard(0)
+        shrd1 = Shard(1)
+        repl = Replicate()
+        part = Partial()
+
+        ws = self.world_size
+        m, n, k = 64 * ws, 48 * ws, 128 * ws
+
+        t1 = torch.randn(m, k, device=self.device_type)
+        t2 = torch.randn(n, k, device=self.device_type)
+
+        for (
+            output_spec,
+            t1_spec,
+            t2_spec,
+            scale1_shape,
+            scale2_shape,
+            scale1_spec,
+            scale2_spec,
+        ) in [
+            # Tensor-wise scaling
+            # Replicated, zero-dim scale
+            (repl, repl, repl, (), (), repl, repl),
+            # Column-parallel, two-dim scale
+            (shrd1, repl, shrd0, (1, 1), (1, 1), repl, repl),
+            # Row-parallel, one-dim scale
+            (part, shrd1, shrd1, (1,), (1,), repl, repl),
+            # Row-wise scaling
+            # Replicated
+            (repl, repl, repl, (m, 1), (n, 1), repl, repl),
+            # Column-parallel
+            (shrd1, repl, shrd0, (m, 1), (n, 1), repl, shrd0),
+            # Row-parallel (which actually ends up doing sub-row-wise scaling)
+            (part, shrd1, shrd1, (m, ws), (n, ws), shrd1, shrd1),
```

Comment:
Just curious: in practice, when are scales of such shapes generated?

Reply:
You can take a look at this PR, which is how we intend to use this sharding strategy in order to make our fp8 recipe compatible with tensor parallelism: https://github.com/facebookresearch/lingua/pull/65. You can see how, in the row-parallel linear case, the matmul is effectively in "split-k" mode, and it's more efficient _and_ more accurate to use a different scaling factor for each of the splits.